luci-base: added datatypes `hostid` and `base64`

danrl opened the pull request:
Signed-off-by: Dan Luedtke <mail@danrl.com>

hnyman commented:
I saw yesterday your comment to the other PR and thought to mention to you that would it maybe make sense to modify that "duid" to be more versatile.

I introduced last week "hexstring" type but did not create a length check as I thought to use combined  "and(hexstring,rangelength(4,4))" for the Wifi options.

But as you are also needing a hex verification of certain length for duid, and there are various r1kh, r1kh options where that could be useful, would it make sense to create "hexlen" instead, and use it like hexlen(16).

Something like:
```
function hexlen(val,len)
        if val and len then
                len = tonumber(len)
                return (#val == len) and (val:match("^[a-fA-F0-9]+$") ~= nil)
        end
        return false
end
```
(not sure what is the most polished formulating)

danrl commented:
Actually, I see no need for `hexlen` since `and(hexstring,rangelength(4,4))` works just fine. I totally overlooked that. However, I do not object introducing it if that makes life easier for some people.

I also introducec `base64` while I am at it. It's needed for this PR: https://github.com/openwrt/luci/pull/1015

danrl commented on modules/luci-base/luasrc/cbi/datatypes.lua:
```diff
@@ -268,17 +277,24 @@ function wepkey(val)
+		return (val:match("^[a-fA-F0-9]+$") ~= nil)
+	end
+	return false
+end
```
spaces to tabs, sorry. Didn't mean to touch your code. At least now it is consistent :)

hnyman commented:
> spaces to tabs

Good catch, and my bad copy-paste from a live router.

But I am not quite sure I understand the "hostid". It checks that a suffix (apparently either 123 or :123:234 or so) is a valid ipv6 host ID? 

It sounds so specific that I think that you should add "ip6" to the name. ip6hostid ?

danrl commented:
> It sounds so specific that I think that you should add "ip6" to the name. ip6hostid ?
`hostid` is the name of the UCI option. I see that `ip6hostid` is a better fit for a datatype name. Changed that. And moved it further up to it's new `ip6*` friends.
